components/speedreader/speedreader_service.cc

```diff
@@ -6,33 +6,94 @@
 #include "brave/components/speedreader/speedreader_service.h"
 
 #include "base/feature_list.h"
+#include "base/metrics/histogram_macros.h"
 #include "brave/components/speedreader/features.h"
 #include "brave/components/speedreader/speedreader_pref_names.h"
+#include "brave/components/weekly_storage/weekly_storage.h"
 #include "components/prefs/pref_registry_simple.h"
 #include "components/prefs/pref_service.h"
 
 namespace speedreader {
 
+namespace {
+
+// Note: append-only enumeration! Never remove any existing values, as this enum
+// is used to bucket a UMA histogram, and removing values breaks that.
+constexpr std::array<uint64_t, 5> kSpeedReaderToggleBuckets{
+    0,   // 0
+    5,   // >0-5
+    10,  // 6-10
+    20,  // 11-20
+    30,  // 21-30
+         // 30+ => bucket 5
+};
+
+constexpr char kSpeedreaderToggleUMAHistogramName[] =
+    "Brave.SpeedReader.ToggleCount";
+
+constexpr char kSpeedreaderEnabledUMAHistogramName[] =
+    "Brave.SpeedReader.Enabled";
+
+void storeTogglesHistogram(uint64_t toggles) {
+  int bucket = 0;
+  for (const auto& bucketUpperBound : kSpeedReaderToggleBuckets) {
+    if (toggles > bucketUpperBound)
+      bucket = bucket + 1;
+  }
+
+  UMA_HISTOGRAM_EXACT_LINEAR(kSpeedreaderToggleUMAHistogramName, bucket, 5);
+}
+
+void storeSpeedReaderEnabledHistogram(PrefService* prefs_,
+                                      bool toggled,
+                                      bool enabledNow) {
+  if (toggled) {
+    WeeklyStorage weekly(prefs_, kSpeedreaderPrefToggleCount);
+    weekly.AddDelta(1);
+    storeTogglesHistogram(weekly.GetWeeklySum());
+  }
+
+  if (toggled && enabledNow) {
+    // Switched on now
+    prefs_->SetBoolean(kSpeedreaderPrefEverEnabled, true);
+    UMA_HISTOGRAM_EXACT_LINEAR(kSpeedreaderEnabledUMAHistogramName, 2, 3);
+  } else {
+    // user has had the feature enabled before pref started getting tracked
```

Comment:
Also per code style commends should look like normal sentences - i.e. a capital letter at the beginning and a dot at the end